crates/host/src/module.rs

```diff
@@ -1,38 +1,136 @@
+use crate::prelude::Instance;
 use holochain_wasmer_common::WasmError;
 use once_cell::sync::Lazy;
+use parking_lot::Mutex;
 use parking_lot::RwLock;
+use plru::MicroCache;
 use std::collections::HashMap;
+// use std::sync::atomic::AtomicU64;
+use bimap::BiMap;
 use std::sync::atomic::AtomicUsize;
 use std::sync::Arc;
 use wasmer::Cranelift;
 use wasmer::Module;
 use wasmer::Store;
 use wasmer::Universal;
 
+pub type CacheKey = [u8; 32];
+pub type PlruKeyMap = BiMap<usize, CacheKey>;
+pub type SerializedModule = Vec<u8>;
+
+trait PlruCache {
+    type Item;
+
+    fn plru_mut(&mut self) -> &mut MicroCache;
+    fn key_map(&self) -> &PlruKeyMap;
+    fn key_map_mut(&mut self) -> &mut PlruKeyMap;
+    fn cache(&self) -> &HashMap<CacheKey, Arc<Self::Item>>;
+    fn cache_mut(&mut self) -> &mut HashMap<CacheKey, Arc<Self::Item>>;
+
+    fn put_item(&mut self, key: CacheKey, item: Self::Item) -> Arc<Self::Item> {
+        let plru_key = self.plru_mut().replace();
+        // if there is something in the cache for this plru slot already drop it.
+        if let Some(stale_key) = self.key_map().get_by_left(&plru_key).cloned() {
+            self.cache_mut().remove(&stale_key);
+        }
+        let arc = Arc::new(item);
```

Comment:
I don't think it makes sense to wrap things in an Arc here because we are caching instances in an arc